### PR TITLE
Get assertions api

### PR DESCRIPTION
**Pull request by vteague**

### CODE REVIEW

File: src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsInProgressValidAPIRequestTests.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    assertTrue(correctSolutionData(320,1.1, 1, response.getBody(),EPS));
+
+    // We expect one assertion with the following data.
+    assertTrue(correctIndexedAPIAssertionData("NEB", 320, 1.1, 0,1, new ArrayList<>(),
+        0.5, response.getBody(),0,EPS));
+
+  }
+
+  /**
+   * Retrieve assertions for a contest that has one NEN assertion (audit in progress).
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsInProgressValidAPIRequestTests.java:121]: Delete blank line between comment and method.

File: src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsInProgressValidAPIRequestTests.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+import org.springframework.test.annotation.DirtiesContext.ClassMode;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Tests for get-assertions endpoint. This class automatically fires up the RAIRE Microservice on a random
+ * port, then runs a series of tests for correct responses to valid requests.
+ * The list of tests is similar to - and in most cases identical to - the GetAssertionsServiceTests.
+ * Note that you have to run the *whole class*. Individual tests do not work separately because they don't
+ * initiate the microservice on their own.
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsInProgressValidAPIRequestTests.java:54]: Delete blank line between comment and class.

File: src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsValidAPIRequestTests.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+public class GetAssertionsValidAPIRequestTests {
+
+  private final static HttpHeaders httpHeaders = new HttpHeaders();
+  private final static String baseURL = "http://localhost:";
+  private final static String getAssertionsEndpoint = "/raire/get-assertions";
+
+  private final static String oneNEBAssertionContest = "One NEB Assertion Contest";
+  private final static String oneNENAssertionContest = "One NEN Assertion Contest";
+  private final static String oneNEBOneNENAssertionContest = "One NEN NEB Assertion Contest";
+
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsValidAPIRequestTests.java:75]: May as well make this comment in the javadoc style for consistency.

File: src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsValidAPIRequestTests.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    // Check the contents of the RaireResults within the RaireSolution.
+    assertTrue(correctSolutionData(320,1.1, 1, response.getBody(),EPS));
+
+    // We expect one assertion with the following data.
+    assertTrue(correctIndexedAPIAssertionData("NEB", 320, 1.1, 0,1, new ArrayList<>(),
+        1.0, response.getBody(),0,EPS));
+
+  }
+
+  /**
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/controller/GetAssertionsValidAPIRequestTests.java:141]: Delete blank line between comment and method.

File: src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java [hunk 5]
```diff
@@ -81,6 +80,9 @@ public class GenerateAssertionsOnKnownTestCases {
   @Autowired
   GenerateAssertionsService generateAssertionsService;
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java:81]: Could replace comment with javadoc style for consistency.

File: src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java [hunk 6]
```diff
@@ -21,10 +21,10 @@ the raire assertion generation engine (https://github.com/DemocracyDevelopers/ra
 package au.org.democracydevelopers.raireservice.service;
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java:23]: Some of the lines of code in this file are too long.

File: src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java [hunk 7]
```diff
@@ -336,16 +326,14 @@ public void simpleContestSingleCountyDoubleBallots() throws RaireServiceExceptio
...
     // There should be one NEN assertion: Alice > Bob if only {Alice,Bob} remain.
     Optional<Assertion> nenMaybeAssertion = assertions.stream()
         .filter(a -> a instanceof NENAssertion).findFirst();
     assertTrue(nenMaybeAssertion.isPresent());
     NENAssertion nenAssertion = (NENAssertion) nenMaybeAssertion.get();
-    assertTrue(correctAssertionData(1, BigDecimal.valueOf(0.1), BigDecimal.valueOf(10),
-        "Alice","Bob", nenAssertion));
+    assertTrue(correctDBAssertionData(1, 0.1, 10, "Alice", "Bob", nenAssertion, EPS));
     assertTrue(correctAssumedContinuing(List.of("Bob","Alice"), nenAssertion));
   }
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java:339]: Typo in method name (insufficient).

File: src/test/java/au/org/democracydevelopers/raireservice/testUtils.java [hunk 8]
```diff
@@ -39,37 +40,144 @@ public class testUtils {
...
+    double retrievedRiskLimit = metadataElement.get("risk_limit").getAsDouble();
+    JsonArray retrievedCandidatesJson = metadataElement.getAsJsonArray("candidates");
+    List<String> retrievedCandidates = new ArrayList<>();
+    for (JsonElement jsonElement : retrievedCandidatesJson) {
+      retrievedCandidates.add(jsonElement.getAsString());
+    }
+
+    return contestName.equals(retrievedContestName)
+        && Math.abs(riskLimit - retrievedRiskLimit) < eps
+        && setsNoDupesEqual(candidates, retrievedCandidates);
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/testUtils.java:69]: Remove blank line between return statement and }.

File: src/test/java/au/org/democracydevelopers/raireservice/testUtils.java [hunk 9]
```diff
@@ -25,10 +25,11 @@ the raire assertion generation engine (https://github.com/DemocracyDevelopers/ra
 import au.org.democracydevelopers.raireservice.persistence.entity.Assertion;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.reflect.TypeToken;
-import java.math.BigDecimal;
+import java.util.ArrayList;
 import java.util.List;
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/testUtils.java:34]: Add class level comment.

Some lines in this file are too long -- use vertical line guide.

File: src/test/java/au/org/democracydevelopers/raireservice/testUtils.java [hunk 10]
```diff
@@ -39,37 +40,144 @@ public class testUtils {
...
+   * as strings.
+   */
+  public static boolean correctIndexedAPIAssertionData(String type, int margin, double difficulty,
+      int winner, int loser, List<Integer> assumedContinuing, double risk, String APIResponseBody, int index, double eps) {
+    JsonObject data =  GSON.fromJson(APIResponseBody, JsonObject.class);
+    String assertion = data.get("solution").getAsJsonObject().get("Ok").getAsJsonObject()
+        .get("assertions").getAsJsonArray().get(index).getAsJsonObject().toString();
+    return correctAPIAssertionData(type, margin, difficulty, winner, loser, assumedContinuing, risk, assertion, eps);
+  }
+
```
Comment (michelleblom) [inline src/test/java/au/org/democracydevelopers/raireservice/testUtils.java:112]: Missing method level comment.

**Review by michelleblom — CHANGES_REQUESTED** [review]

File: src/main/java/au/org/democracydevelopers/raireservice/json/RaireResultMixIn.java [hunk 11]
```diff
@@ -20,15 +20,18 @@ the raire assertion generation engine (https://github.com/DemocracyDevelopers/ra
```
[the remark this answers is not in the document]
Comment (vteague) [reply src/main/java/au/org/democracydevelopers/raireservice/json/RaireResultMixIn.java:20]: Now moved into the response directory.

File: src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java (hunk 6 above)
Comment (vteague) [reply src/test/java/au/org/democracydevelopers/raireservice/service/GenerateAssertionsOnKnownTestCases.java:23]: Cut to 110.

**Review by michelleblom — APPROVED** [review]

**Pull request merged**